AK+LibJS: Handle NaN-boxing pointers on AArch64

BertalanD opened the pull request:
JS::Value stores 48 bit pointers to separately allocated objects in its payload. On x86-64, canonical addresses have their top 16 bits set to the same value as bit 47, effectively meaning that the value has to be sign-extended to get the pointer. AArch64, however, expects the topmost bits to be all zeros.

This commit gates sign extension behind `#if ARCH(X86_64)`, and adds an `#error` for unsupported architectures, so that we do not forget to choose the correct behavior when porting to a new architecture.

Fixes #15290
Fixes SerenityOS/ladybird#56

---
~~This is a draft for now, as I haven't  tested this yet on my AArch64 Ubuntu VM.~~

Ladybird runs and `test-js` now passes on AArch64 linux!

ADKaster commented on Userland/Libraries/LibJS/Runtime/Value.h:
```diff
@@ -398,6 +398,24 @@ class Value {
     template<typename... Args>
     [[nodiscard]] ALWAYS_INLINE ThrowCompletionOr<Value> invoke(VM&, PropertyKey const& property_key, Args... args);
 
+    static FlatPtr extract_pointer_bits(uint64_t encoded)
```
`uint64_t`?

One of the call sites passes FlatPtr here, which is 32 bits sometimes (Heap.cpp).

At any rate it should be u64 since Value::m_bits::encoded is u64

ADKaster replied:
Side note can this be a constexpr method?

davidot replied:
This method shouldn't be called from Heap when when FlatPtr is 32 bits.
But yes it should be `u64` and can be constexpr

ADKaster commented on Userland/Libraries/LibJS/Runtime/Value.h:
```diff
@@ -398,6 +398,24 @@ class Value {
+#elif ARCH(X86_64)
+        // For x86_64 the top 16 bits should be sign extending the "real" top bit (47th).
+        // So first shift the top 16 bits away then using the right shift it sign extends the top 16 bits.
+        return static_cast<FlatPtr>(((i64)(encoded << 16)) >> 16);
```
Static cast and c style cast in the same expression? :(

davidot reviewed (commented):
Looks fine to me, I agree with nits about constexpr and casting but not a big deal.

You might also want to update https://github.com/SerenityOS/serenity/blob/master/Tests/LibJS/test-value-js.cpp#L44 as these tests will probably fail on ARM.
But since we don't have an ARM CI currently it's not a huge deal to do this later. 
To make sure we still test this issuewe might need to change more so I will have a look at it later.

BertalanD commented:
> You might also want to update https://github.com/SerenityOS/serenity/blob/master/Tests/LibJS/test-value-js.cpp#L44 as these tests will probably fail on ARM.

I updated the test to use the new `JS::Value::extract_pointer_bits` function and added tests for the arm64 behavior. A quick `rg "<< 16.* >> 16"` search shows that there are no more manual sign extensions left, is that correct?

I fixed all the style issues too.

davidot reviewed (approved):
Yeah looks good, the Heap should be the only other spot.

Thanks for the tests, I think having an ifdef there is fine since it clearly shows why there is a difference.
